tests/store/test_sqlalchemy_store.py

```diff
@@ -398,6 +399,26 @@ def test_log_metric(self):
 
         self.assertTrue(found)
 
+    def test_log_metric_allows_multiple_values_at_same_timestamp_and_read_returns_max_value(self):
+        run = self._run_factory()
+
+        self.session.commit()
+
+        num_metrics_for_timestamp = 10
+        metric_key = "duplicate_timestamp_metric"
+        metric_vals = [random.random() for _ in range(num_metrics_for_timestamp)]
+        metric_timestamp = time.time()
+        for metric_val in metric_vals:
+            metric = entities.Metric(metric_key, metric_val, metric_timestamp)
+            self.store.log_metric(run.run_uuid, metric)
+
+        timestamp_duplicate_metrics = [
+            metric for metric in self.store.get_run(run.run_uuid).data.metrics
+            if metric.key == metric_key
+        ]
+        assert len(timestamp_duplicate_metrics) == 1
+        assert timestamp_duplicate_metrics[0].value == max(metric_vals)
+
     def test_log_metric_uniqueness(self):
```

Comment:
aaah!! There was a testcase for this... aaha!!
I would make it more robust like so...

```
        timestamp = time.time()
        metric = entities.Metric('blahmetric', 100.0, int(timestamp))
        metric2 = entities.Metric('blahmetric', 1.02, int(timestamp))
        self.store.log_metric(run.run_uuid, metric)
        # no assert
        self.store.log_metric(run.run_uuid, metric2)

        self.assert(set(self.store.get_metric_history('blahmetric')) == set([100.0, 1.02]))
        self.assert(self.store.get_metric('blahmetric') == 100.0)
```

Reply:
Thanks for including code in the suggestion! I like the idea of using values that are explicitly different, as well as checking `get_metric_history()`. I've made those changes. Unfortunately, `get_metric()` is not part of the store API, so it seems that we still need to rely on filtering `run.data.metrics`.